src/storage/v2/disk/storage.cpp

```diff
@@ -774,98 +774,278 @@ std::optional<VertexAccessor> DiskStorage::DiskAccessor::FindVertex(storage::Gid
+        std::unique_lock<utils::SpinLock> guard_vertex(vertex_ptr->lock, std::defer_lock);
+        std::unique_lock<utils::SpinLock> guard_opposing_vertex(opposing_vertex->lock, std::defer_lock);
+
+        // Obtain the locks by `gid` order to avoid lock cycles.
+        if (vertex_ptr->gid < opposing_vertex->gid) {
+          guard_vertex.lock();
+          guard_opposing_vertex.lock();
+        } else if (vertex_ptr->gid > opposing_vertex->gid) {
+          guard_opposing_vertex.lock();
+          guard_vertex.lock();
+        } else {
+          // The vertices are the same vertex, only lock one.
+          guard_vertex.lock();
+        }
+
+        if (!PrepareForWrite(&transaction_, vertex_ptr)) return Error::SERIALIZATION_ERROR;
+        MG_ASSERT(!vertex_ptr->deleted, "Invalid database state!");
+
+        if (opposing_vertex != vertex_ptr) {
+          if (!PrepareForWrite(&transaction_, opposing_vertex)) return Error::SERIALIZATION_ERROR;
+          MG_ASSERT(!opposing_vertex->deleted, "Invalid database state!");
+        }
+
+        edges_collection->pop_back();
+
+        auto *from_vertex = reverse_vertex_order ? vertex_ptr : opposing_vertex;
+        auto *to_vertex = reverse_vertex_order ? opposing_vertex : vertex_ptr;
+        mark_edge_as_deleted_func(edge_ref, edge_type, from_vertex, to_vertex);
 
-  // We need to check again for serialization errors because we unlocked the
-  // vertex. Some other transaction could have modified the vertex in the
-  // meantime if we didn't have any edges to delete.
+        CreateAndLinkDelta(&transaction_, vertex_ptr, delta, edge_type, opposing_vertex, edge_ref);
+      }
 
-  if (!PrepareForWrite(&transaction_, vertex_ptr)) return Error::SERIALIZATION_ERROR;
+      return std::make_optional<ReturnType>();
+    };
 
-  MG_ASSERT(!vertex_ptr->deleted, "Invalid database state!");
+    // delete the in and out edges from the nodes we want to delete
+    for (auto *vertex_ptr : nodes_to_delete) {
+      auto maybe_error = clear_edges(vertex_ptr, &vertex_ptr->in_edges, Delta::AddInEdgeTag(), false);
+      if (maybe_error.HasError()) {
+        return maybe_error;
+      }
 
-  CreateAndLinkDelta(&transaction_, vertex_ptr, Delta::RecreateObjectTag());
-  vertex_ptr->deleted = true;
-  vertices_to_delete_.emplace_back(utils::SerializeIdType(vertex_ptr->gid), utils::SerializeVertex(*vertex_ptr));
+      maybe_error = clear_edges(vertex_ptr, &vertex_ptr->out_edges, Delta::AddOutEdgeTag(), true);
+      if (maybe_error.HasError()) {
+        return maybe_error;
+      }
+    }
+  }
 
-  return std::make_optional<ReturnType>(
-      VertexAccessor{vertex_ptr, &transaction_, &storage_->indices_, &storage_->constraints_, config_, true},
-      std::move(deleted_edges));
+  // Detach nodes on the other end, which don't need deletion, by passing once through their vectors
+  {
+    auto detach_non_deletable_nodes = [this, &mark_edge_as_deleted_func](auto *vertex_ptr, auto *edges_collection,
```

Comment:
so the name is `detach_non_deletable_nodes` but here `detach` can also be `false`? Even if it's not wrong, it misguides